python/ray/data/_internal/datasource/clickhouse_datasource.py

```diff
@@ -0,0 +1,180 @@
+        """
+        Create read tasks for the ClickHouse query.
+
+        Args:
+            parallelism: The desired number of partitions to read the data into.
+
+        Returns:
+            A list of read tasks to be executed.
+        """
+        num_rows_total = self._get_estimate("count")
+        if num_rows_total == 0 or num_rows_total is None:
+            return []
+        parallelism = min(
+            parallelism, math.ceil(num_rows_total / self.MIN_ROWS_PER_READ_TASK)
+        )
+        num_rows_per_block = num_rows_total // parallelism
+        num_blocks_with_extra_row = num_rows_total % parallelism
+        sample_block_accessor = BlockAccessor.for_block(self._get_sample_block())
+        estimated_size_bytes_per_row = math.ceil(
+            sample_block_accessor.size_bytes() / sample_block_accessor.num_rows()
+        )
+        sample_block_schema = sample_block_accessor.schema()
+        read_tasks = []
+        offset = 0
+        for i in range(parallelism):
+            num_rows = num_rows_per_block
+            if i < num_blocks_with_extra_row:
+                num_rows += 1
+            read_fn = self._create_read_fn(num_rows, offset)
+            metadata = BlockMetadata(
+                num_rows=num_rows,
+                size_bytes=estimated_size_bytes_per_row * num_rows,
+                schema=sample_block_schema,
+                input_files=None,
+                exec_stats=None,
+            )
+            read_tasks.append(ReadTask(read_fn, metadata))
+            offset += num_rows
+        return read_tasks
+
+    def _get_estimate(self, query_type: str) -> Optional[int]:
+        client = self._init_client()
+        try:
+            result = client.query(self._estimates[query_type])
+            if result and len(result.result_rows) > 0:
+                estimate = result.result_rows[0][0]
+                return int(estimate) if estimate is not None else None
+        except Exception as e:
+            logger.warning(f"Failed to estimate query {query_type}: {e}")
+        finally:
+            client.close()
+        return None
+
+    def _get_sample_block(self) -> Block:
+        import pyarrow as pa
+
+        client = self._init_client()
+        try:
+            query = f"SELECT * FROM ({self._query}) LIMIT {self.NUM_SAMPLE_ROWS}"
+            with client.query_arrow_stream(query) as stream:
+                record_batches = list(stream)  # Collect all record batches
+            return pa.Table.from_batches(record_batches)
+        except Exception as e:
+            logger.warning(f"Failed to get sample block: {e}")
+        finally:
+            client.close()
+
+    def _create_read_fn(
+        self, num_rows: int, offset: int
+    ) -> Callable[[], Iterable[Block]]:
+        def read_fn() -> Iterable[Block]:
+            import pyarrow as pa
+
+            client = self._init_client()
+            query = f"SELECT * FROM ({self._query}) LIMIT {num_rows} OFFSET {offset}"
```

Comment:
@jecsand838 does Clickhouse guarantee to always traverse the rows in the same order?

If that's not the case we'd have to revert back to reading the whole dataset in 1 task unfortunately (we've been recently bitten by this for SQL datasource)

Reply:
You have to explicitly define an `ORDER BY` clause to guarantee order. Let me know if that's acceptable (I can update the docs) or if we need to revert still.

Reply:
Actually there maybe a way to get the behavior we want in ClickHouse. I'll look into it and get back.

Reply:
So, what we can do is following:

1. If users want parallelism we will require order-by to be specified
2. If they don't care/can't provider order-by then will be read as 1 task

Does that make sense?

Reply:
@alexeykudinkin That makes sense. I went ahead and attempted those changes. Let me know what you think.